### PR TITLE
refactor: add `Configurable` trait

### PR DESCRIPTION
**Pull request by zyy17**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Add the `Configurable` trait to make configuration more general.

1. Add the new module `error.rs` and `config.rs` in `common/config`. Most of the code are from `cmd/`:

   - The `load_layer_options()` is from the original `cmd/options.rs`;
   - The errors are from `cmd/error.rs`;
   
3. Add `merge_with_cli_options()` to make `load_options()` more simple;
4. Implement `Configurable` trait and changed some tests for `{DatanodeOptions, FrontendOptions, MetasrvOptions, StandaloneOptions}`;

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/cmd/src/error.rs [hunk 1]
```diff
@@ -210,7 +209,7 @@ pub enum Error {
     #[snafu(display("Failed to load layered config"))]
     LoadLayeredConfig {
         #[snafu(source)]
-        error: ConfigError,
+        source: Box<common_config::error::Error>,
```
Comment (evenyag) [inline src/cmd/src/error.rs:212]: https://docs.rs/snafu/latest/snafu/derive.Snafu.html#transforming-the-source
This should get rid of `map_err(Box::new)`
```suggestion
        #[snafu(source(from(common_config::error::Error, Box::new)))]
        source: Box<common_config::error::Error>,
```

File: src/datanode/src/error.rs [hunk 2]
```diff
@@ -359,6 +359,13 @@ pub enum Error {
         #[snafu(implicit)]
         location: Location,
     },
+
+    #[snafu(display("Failed to serialize options to TOML"))]
+    TomlFormat {
+        #[snafu(implicit)]
+        location: Location,
+        source: Box<common_config::error::Error>,
```
Comment (evenyag) [inline src/datanode/src/error.rs:367]: Try https://docs.rs/snafu/latest/snafu/derive.Snafu.html#transforming-the-source

File: src/frontend/src/error.rs [hunk 3]
```diff
@@ -343,8 +343,9 @@ pub enum Error {
 
     #[snafu(display("Failed to serialize options to TOML"))]
     TomlFormat {
-        #[snafu(source)]
-        error: toml::ser::Error,
+        #[snafu(implicit)]
+        location: Location,
+        source: Box<common_config::error::Error>,
```
Comment (evenyag) [inline src/frontend/src/error.rs:348]: ditto

File: src/common/config/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            layered_config = layered_config.add_source(File::new(config_file, FileFormat::Toml));
+        }
+
+        let opts = layered_config
+            .build()
+            .context(LoadLayeredConfigSnafu)?
+            .try_deserialize()
+            .context(LoadLayeredConfigSnafu)?;
+
+        Ok(opts)
```
Comment (MichaelScofield) [inline src/common/config/src/config.rs:81]: Try this:
```suggestion
        layered_config
            .build()
            .and_then(|x| x.try_deserialize())
            .context(LoadLayeredConfigSnafu)
```

**Review by MichaelScofield — APPROVED** [review]

File: src/common/config/src/config.rs (hunk 4 above)
Comment (zyy17) [reply src/common/config/src/config.rs:81]: TIL

**Comment by zyy17** [comment]
@evenyag PTAL

**Review by evenyag — APPROVED** [review]

**Pull request merged**